### DISCUSSION TITLE
how to ignore empty inputs if field isn't required

### DISCUSSION THREAD

**Discussion by weizwx** [fastapi/fastapi#9697] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [x] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [ ] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [ ] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Body

app = FastAPI()

@app.post('/test')
def _test(
  _id: int | None = Body(alias='id', default=None),
  amount: int = Body(default=0)
):
    return {
      'id': _id,
      'amount': amount
    }
```

### Description

Currently, if a form field is not required and leaves it blank, it can raise an error if an empty value string ("") isn't valid, i want to return the default value when the frontend is not entered.

`curl -X 'POST' \
  'http://localhost:8000/test' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "id": "",
  "amount": ""
}'`

*It return a JSON with 
`{
  "errors": [
    {
      "loc": [
        "body",
        "id"
      ],
      "msg": "value is not a valid integer",
      "type": "type_error.integer"
    },
    {
      "loc": [
        "body",
        "amount"
      ],
      "msg": "value is not a valid integer",
      "type": "type_error.integer"
    }
  ],
  "body": {
    "id": "",
    "amount": ""
  }
}`

*But i expected it to return
`
{
  "id": null,
  "amount": 0
}
`

### Operating System

Linux

### Operating System Details

ubuntu22.04

### FastAPI Version

0.97.0

### Python Version

Python 3.10.11

**Comment by abhint** [comment]
```py
from fastapi import FastAPI, Body, Depends
from pydantic import BaseModel, Field, validator
from typing import Optional

class TestInput(BaseModel):
    id: Optional[int] = None
    amount: int = 0

    @validator('id', pre=True)
    def validate_empty_id(cls, value):
        if value == '':
            return None
        return value

app = FastAPI()

@app.post('/test')
def _test(input_data: TestInput = Depends(TestInput.as_form)):
    return {
        'id': input_data.id,
        'amount': input_data.amount
    }
```
You will receive the expected response:

```json
{
  "id": null,
  "amount": 0
}

```

**Reply by weizwx** [reply]
Thank you for your reply. But there may be many similar parameters, such as datetime type parameters.It would be nice if there was a more convenient way

**Reply by abhint** [reply]
```py
from typing import TypeVar, Any, Callable
from fastapi import Body
from pydantic import validator

T = TypeVar('T')


def handle_empty_input(default_value: T, validator_func: Callable[[Any], T]):
    def wrapper(field: T = Body(...)):
        return field if field != "" else default_value

    return validator_func(wrapper, pre=True, always=True)


class TestInput:
    id: int = None
    amount: int = 0
    created_at: datetime = None

    check_empty = handle_empty_input(None, validator)


@app.post('/test')
def _test(input_data: TestInput = Body(...)):
    return {
        'id': input_data.id,
        'amount': input_data.amount,
        'created_at': input_data.created_at
    }

```

**Comment by wu-clan** [comment]
You should use type passing correctly

```py
from fastapi import FastAPI, Body

app = FastAPI()

@app.post('/test')
def _test(
  _id: int | str | None = Body(alias='id', default=None),
  amount: int | str = Body(default=0)
):
    return {
      'id': _id,
      'amount': amount
    }
```

In this case, id is allowed to be integer, string, or null, and amount is allowed to be integer, string.

The type of return result is determined by the parameter type.

'{"id": null, "amount": ""}' -> {"id": null, "amount": ""}
'{"id": null, "amount": 0}' -> {"id": null, "amount": 0}
'{"id": "", "amount": ""}' -> {"id": "", "amount": ""}
...

Otherwise your practice type validation becomes meaningless

**Comment by YuriiMotov** [comment, marked as answer]
I think this is the way to go:

```python
from typing import Annotated, TypeAlias, TypeVar
from fastapi import FastAPI, Form
from fastapi.testclient import TestClient
from pydantic import BaseModel, BeforeValidator

app = FastAPI()


T = TypeVar("T")

EmptyToNone: TypeAlias = Annotated[
    T, BeforeValidator(lambda v: None if (v == "") else v)
]


@app.get("/test_query")
async def query(param: EmptyToNone[int | None] = None):
    return {"param": param}


class Params(BaseModel):
    v: EmptyToNone[int | None] = None


@app.post("/test_form_model")
async def form(params: Params = Form()):
    return {"params": params}


# =====================================================================
# Tests


client = TestClient(app)


def test_query_param_none():
    # No param provided
    response = client.get("/test_query")
    assert response.status_code == 200
    assert response.json() == {"param": None}


def test_query_param_int():
    # Valid int param
    response = client.get("/test_query", params={"param": 123})
    assert response.status_code == 200
    assert response.json() == {"param": 123}


def test_query_param_empty_string():
    # Empty string should be converted to None
    response = client.get("/test_query", params={"param": ""})
    assert response.status_code == 200
    assert response.json() == {"param": None}


def test_query_param_invalid():
    # Invalid int param
    response = client.get("/test_query", params={"param": "abc"})
    assert response.status_code == 422


def test_form_model_none():
    # No form data
    response = client.post("/test_form_model", data={})
    assert response.status_code == 200
    assert response.json() == {"params": {"v": None}}


def test_form_model_int():
    # Valid int in form
    response = client.post("/test_form_model", data={"v": "42"})
    assert response.status_code == 200
    assert response.json() == {"params": {"v": 42}}


def test_form_model_empty_string():
    # Empty string in form should be converted to None
    response = client.post("/test_form_model", data={"v": ""})
    assert response.status_code == 200
    assert response.json() == {"params": {"v": None}}


def test_form_model_invalid():
    # Invalid int in form
    response = client.post("/test_form_model", data={"v": "foo"})
    assert response.status_code == 422


```
